### ISSUE TITLE
set encoding in calculateMD5hash()

### ISSUE THREAD

**Issue by okomarov** [matlab2tikz/matlab2tikz#775]
I came across http://stackoverflow.com/a/415971/2180721 which warns about setting the enconding of the input stream.

This could solve some unreliable hashes if we set a common encoding, e.g. `UTF-8`

**Comment by PeterPablo** [comment]
Do you have a suspicion that the encoding is cause of some unreliable hashes? It hasn't crossed my path up to now.

**Comment by egeerardyn** [comment]
I don't think it's currently a problem, but better safe than sorry.
When adapting the code, we should really provide a link back to that page.

**Comment by okomarov** [comment]
It might. For instance when I load ACID.m, it always tries this:

![capture](https://cloud.githubusercontent.com/assets/3731173/9936687/16270e02-5d55-11e5-856d-971bd7a85716.PNG)

**Comment by egeerardyn** [comment]
Hmm, if I look at the example from StackOverflow, that is specifically for `String` objects. We currently don't pass any `java.String` for that purpose, but rather calculate the hash directly from an input stream:
 
```matlab
fis = java.io.FileInputStream(java.io.File(absoluteFilename));
MD5 = java.security.MessageDigest.getInstance('MD5');
dis = java.security.DigestInputStream(fis, MD5);

while(dis.read() ~= -1), end; % read the whole file

hash = reshape(dec2hex(typecast(MD5.digest(),'uint8')).', 1, 32);
```

So, as far as I understand it, the file is not read out to a `String` object but rather to a [binary](http://docs.oracle.com/javase/7/docs/api/java/io/FileInputStream.html) stream of data which is [digested]((http://docs.oracle.com/javase/7/docs/api/java/security/DigestInputStream.html)). As such, encoding doesn't come into play here, since there is no need to map bit/byte sequences (e.g. `0x3141`) to characters (the characters `1A` in plain ASCII). When calculating the hash on a `String`, the inverse has to happen: you have to translate characters back to a (specifically encoded) sequence of bytes.

What does matter, however, is in what encoding we write out the files in. We only set it in MATLAB, and the default is `UTF-8`. So all should be fine for MATLAB. For Octave, we don't get access to encodings as far as I know, so there is probably nothing we can do.

Anyway, I'm by no means an expert on [character encoding](http://www.joelonsoftware.com/articles/Unicode.html) (I do always have to twist my mind a bit to reason about it, but I guess many people have that when the subject is coding/decoding of anything).
But if my reasoning is flawed, please, do speak up.

**TL;DR:** This can't be an issue on MATLAB. Let's close this topic.

**Comment by okomarov** [comment]
:+1: Thanks for looking into that.

**Comment by egeerardyn** [comment]
You're welcome! Thank you for scrutinizing that part.

Unfortunately, I don't know what your SmartGit is thinking over there. You might want to contact the developers, since it seems like an encoding problem indeed.

**Comment by okomarov** [comment]
It's not smartgit. I suspect it's the encoding of the .m file:

![capture](https://cloud.githubusercontent.com/assets/3731173/9964871/a8d9a734-5e2a-11e5-8590-dfbf92225fc5.PNG)

**Comment by egeerardyn** [comment]
Files don't store the encoding usually. The program you load it with is supposed to interpret it in some way.

And personally, I don't trust TMW with encoding, as I know that on Mac, the encoding is not right by default either. Not that it matters, I have added `feature('DefaultCharacterSet', 'UTF8');` to my MATLAB without any effect.

Just as an example: Nico's name gets garbled still. MATLAB vs. Sublime Text 3 vs. GitHub/Safari. Only MATLAB seems to get it wrong (despite me setting the encoding to UTF8).
<img width="640" alt="screen shot 2015-09-18 at 18 56 10" src="https://cloud.githubusercontent.com/assets/177360/9965646/cf6dc3f0-5e37-11e5-9fab-cdd458553777.png">
<img width="640" alt="screen shot 2015-09-18 at 18 56 26" src="https://cloud.githubusercontent.com/assets/177360/9965648/d2c0e122-5e37-11e5-8a93-63e17a531932.png">
<img width="640" alt="screen shot 2015-09-18 at 18 57 24" src="https://cloud.githubusercontent.com/assets/177360/9965633/b8d8b7c6-5e37-11e5-96cb-011db244b295.png">

**Comment by okomarov** [comment]
Hahaha, storing encoding within a file could be a good way to troll...
 
Yep, looked around, there is no way to set encoding on Matlab's Editor. What a bummer.

**Comment by egeerardyn** [comment]
@okomarov It does happen. See [Joel on Software](http://www.joelonsoftware.com/articles/Unicode.html): in HTML this is a usual practice which relies on the fact that the first nibble is common among most encodings. But I completely agree, it's a f*ing bootstrapping.

And in any file, either you have to pick one by convention, store it in the file or live in eternal ambiguity. All approaches are poor ones.

**Comment by okomarov** [comment]
Yep I already read your link. Good read, thanks.
